pkg/controller/admissionchecks/multikueue/multikueuecluster.go

```diff
@@ -0,0 +1,345 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+)
+
+type clientWithWatchBuilder func(config []byte, options client.Options) (client.WithWatch, error)
+
+type remoteClient struct {
+	localClient client.Client
+	client      client.WithWatch
+	wlUpdateCh  chan<- event.GenericEvent
+	watchCancel func()
+	kubeconfig  []byte
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	builderOverride clientWithWatchBuilder
+}
+
+func newRemoteClient(localClient client.Client, wlUpdateCh chan<- event.GenericEvent) *remoteClient {
+	rc := &remoteClient{
+		wlUpdateCh:  wlUpdateCh,
+		localClient: localClient,
+	}
+	return rc
+}
+
+func newClientWithWatch(kubeconfig []byte, options client.Options) (client.WithWatch, error) {
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	return client.NewWithWatch(restConfig, options)
+}
+
+// setConfig - will try to recreate the k8s client and restart watching if the new config is different than
+// the one currently used.
+func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) error {
+	if equality.Semantic.DeepEqual(kubeconfig, rc.kubeconfig) {
+		return nil
+	}
+
+	if rc.watchCancel != nil {
+		rc.watchCancel()
+		rc.watchCancel = nil
+	}
+
+	builder := newClientWithWatch
+	if rc.builderOverride != nil {
+		builder = rc.builderOverride
+	}
+	remoteClient, err := builder(kubeconfig, client.Options{Scheme: rc.localClient.Scheme()})
+	if err != nil {
+		return err
+	}
+	rc.client = remoteClient
+
+	newWatcher, err := rc.client.Watch(watchCtx, &kueue.WorkloadList{})
+	if err != nil {
+		return nil
+	}
+
+	go func() {
+		for r := range newWatcher.ResultChan() {
+			rc.queueWorkloadEvent(watchCtx, r)
+		}
+	}()
+
+	rc.watchCancel = newWatcher.Stop
+	rc.kubeconfig = kubeconfig
+	return nil
+}
+
+func (rc *remoteClient) queueWorkloadEvent(ctx context.Context, ev watch.Event) {
+	wl, isWl := ev.Object.(*kueue.Workload)
+	if !isWl {
+		return
+	}
+
+	localWl := &kueue.Workload{}
+	if err := rc.localClient.Get(ctx, client.ObjectKeyFromObject(wl), localWl); err == nil {
+		if !apierrors.IsNotFound(err) {
+			ctrl.LoggerFrom(ctx).Error(err, "reading local workload")
+		}
+		rc.wlUpdateCh <- event.GenericEvent{Object: localWl}
+	}
+}
+
+// clustersReconciler implements the reconciler for all MultiKueueClusters.
+// Its main task being to maintain the list of remote clients associated to each MultiKueueCluster.
+type clustersReconciler struct {
+	client          client.Client
+	configNamespace string
+
+	lock sync.RWMutex
+	// The list of remote clients, indexed by the cluster name.
+	clients    map[string]*remoteClient
```

Comment:
```suggestion
	remoteClients    map[string]*remoteClient
```

To clarify the differences between both clients in this structure.

Or

```golang
type clusterReconciler struct {
	localClient client.Client
...
```